src/events/attributes.rs

```diff
@@ -572,6 +474,427 @@ impl std::error::Error for AttrError {}
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
+/// A struct representing a key/value XML or HTML [attribute].
+///
+/// [attribute]: https://www.w3.org/TR/xml11/#NT-Attribute
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Attr<T> {
+    /// Attribute with value enclosed in double quotes (`"`). Attribute key and
+    /// value provided. This is a canonical XML-style attribute
+    DoubleQ(T, T),
+    /// Attribute with value enclosed in single quotes (`'`). Attribute key and
+    /// value provided. This is an XML-style attribute
+    SingleQ(T, T),
+    /// Attribute with value not enclosed in quotes. Attribute key and value
+    /// provided. This is HTML-style attribute, it can be returned in HTML-mode
+    /// parsing only. In an XML mode [`AttrError::UnquotedValue`] will be raised
+    /// instead.
+    ///
+    /// Attribute value can be invalid according to the [HTML specification],
+    /// in particular, it can contain `"`, `'`, `=`, `<`, and <code>&#96;</code>
+    /// characters. The absence of the `>` character is nevertheless guaranteed,
+    /// since the parser extracts [events] based on them even before the start
+    /// of parsing attributes.
+    ///
+    /// [HTML specification]: https://html.spec.whatwg.org/#unquoted
+    /// [events]: crate::events::Event::Start
+    Unquoted(T, T),
+    /// Attribute without value. Attribute key provided. This is HTML-style attribute,
+    /// it can be returned only in HTML-mode parsing only. In XML mode
+    /// [`AttrError::ExpectedEq`] will be returned instead
+    Empty(T),
+}
+
+impl<T> Attr<T> {
+    /// Maps an `Attr<T>` to `Attr<U>` by applying a function to a contained key and value.
+    #[inline]
+    pub fn map<U, F>(self, mut f: F) -> Attr<U>
```

Comment:
(Regarding this whole section of `Attr` code, not this specific part)

One one hand, I completely see where you're coming from with all of this - the enum makes invalid states unrepresentable.

On the other hand, it would add a lot more branches to the generated code, and also a fair amount of bloat when it gets inlined everywhere that a dev uses any of these methods, and it largely feels like a detail few users would care too much about, even ones parsing HTML.

As far as I can tell, HTML doesn't have any attributes that are required to not have a value - attributes with no value, and attributes with an empty value are treated identically.

> A number of attributes are boolean attributes. The presence of a boolean attribute on an element represents the true value, and the absence of the attribute represents the false value.
> 
> If the attribute is present, its value must either be the empty string or a value that is an [ASCII case-insensitive](https://infra.spec.whatwg.org/#ascii-case-insensitive) match for the attribute's canonical name, with no leading or trailing whitespace.

https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes

So my recommendation would be to consider leaving it the way it was before (as a struct with a key and value field), but maybe include a non-data-carrying enum to denote the type if it's still information you would like to provide.

Reply:
That benchmarks shows only small 5 us regression (from ~75 to ~80 with checks, from ~44 to ~49 without checks, from ~155 to ~159 for `try_get_attribute`) in parsing which, I think, not a big pay for an additional info, that could be useful:
https://github.com/Mingun/fast-xml/blob/6a3fd90fa4b6b080ce7f0bd09499c7123e1cc5d9/benches/bench.rs#L256-L333

Besides, I'll also have a next PR with `Span` support and I want to expose the information from where that or another piece of data come. That is dependent of the attribute shape and useful for reporting errors.

bq.  I wouldn't have a problem with a third field to denote unquoted, empty, single-quote, double-quote though. It's just not important enough to treat like `Option` or `Result` and force dealing with specific variants.

That wouldn't be idiomatic Rust. Usually you will use `.key()` and `.value()` methods, so you hardly likely will deal with the concrete variant. Moreover, I've introduced a new `Attr` type instead of modifying `Attribute` to allow users to not deal with attribute shape when it do not matter. There will only one conversion from `Attr` to `Attribute` that just moves shape erasing from internals of iterator to a user-defined place. That should not bloat code well and likely will be optimized by the compiler if you did that just after getting the attribute from the iterator.

Reply:
>That wouldn't be idiomatic Rust. 

There are cases where it actually is idiomatic, or at least a common pattern.  `[std::io::ErrorKind](https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html)` is used because all variants have the same signature.   `[nom::error::ErrorKind](https://docs.rs/nom/7.1.1/nom/error/enum.ErrorKind.html)` is the same.

The problem with going purely off the benchmarks here is that depending on the consistency of the intput, the branch predictor is going to get it right most of the time, so it's not going to be *that* expensive, but it's very dependent on the input, and how much the extra bloat hurts depends a lot on what code the user is writing and how much cache their CPU has.  It's just a lot of extra stuff to put in such a hot path.  

Maybe it could look like this?

```Rust
struct Attr {
    key: ...,
    value: ...,
    shape: AttrShape,
}

enum AttrShape {
    Unqoted(..span stuff..),
    DoubleQuoted(..span stuff..),
    SingleQuoted(..span stuff..),
    Empty(..span stuff..),
}
```

Reply:
Actually, that what you suggesting, is implemented now:
- your `AttrShape` is `Attr<Range<usize>>` -- that is the type that returned by `IterState::next()`
- your `Attr` is the future `Attribute`

Because we in any case need to get spans for `key` and `value`, and because `value` can be missed, and because I want to keep only one parser implementation and not copy it for different types, the current `IterState` design seems to me the almost only possible (the other way is return `(Range<usize>, Option<Range<usize>>)` instead but I don't think that it gives any benefits while reducing code clarity).

I see that your point in possible code bloat and/or inconveniences to use `match` constantly, but that is impossible to trigger from the user code now, because the only way to create an `Attr` instance in the user code is to create it by yourself. `IterState` is a private type. The only place where conversion from `Attr` to `Attribute` is performed now is inside `Attributes::next()`:
https://github.com/Mingun/fast-xml/blob/eaebf458abd555983242f2b6c00d9728b7e5a971/src/events/attributes.rs#L335-L346

I still believe that compiler is able to optimize all right there. Maybe that place seems too generic, because I hide conversion under `map() + Into<Attribute>`, but it is fairly straight-ford